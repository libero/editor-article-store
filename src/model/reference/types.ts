export type ReferenceType =
  | 'journal'
  | 'periodical'
  | 'book'
  | 'report'
  | 'data'
  | 'web'
  | 'preprint'
  | 'software'
  | 'confproc'
  | 'thesis'
  | 'patent';

export type ReferenceContributor =
  | {
  firstName?: string;
  lastName?: string;
}
  | {
  groupName?: string;
};