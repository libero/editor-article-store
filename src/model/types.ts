type Primitives = number | string | boolean | null | undefined;

export type JSONObject = {
    [k: string]: Primitives | JSONObject | Array<Primitives | JSONObject>;
};
