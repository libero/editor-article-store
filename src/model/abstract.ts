import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import jsdom from "jsdom";

import { makeSchemaFromConfig } from './utils';
import * as abstractConfig from './config/abstract.config';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = new jsdom.JSDOM('').window.document;

  if (content) {
    xmlContentDocument.body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = new jsdom.JSDOM('').window.document;

  if (content) {
    xmlContentDocument.body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema
  });
}
