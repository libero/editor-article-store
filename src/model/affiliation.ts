import { BackmatterEntity } from "./backmatter-entity";
import { JSONObject } from "./types";

export class Affiliation extends BackmatterEntity {
  label: string | undefined;
  institution: { name: string; } | undefined;
  address: { city: string; } | undefined;
  country: string | undefined;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }

  protected fromXML(_xml: Element): void {}

  protected fromJSON(json: JSONObject): void {
    this._id = (json._id as string) || this.id;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }
}