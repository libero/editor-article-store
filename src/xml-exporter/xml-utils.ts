import xmldom from "xmldom";
import nwmatcher from "nwmatcher";

export function parseXML(xml: string) {

  const xmlDoc = new xmldom.DOMParser().parseFromString(xml, 'text/xml');

  xmlDoc.constructor.prototype.querySelector = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).first(selectors, this);
  };

  xmlDoc.constructor.prototype.querySelectorAll = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).select(selectors, this);
  };

  const Element = xmlDoc.createElement('a').constructor;

  Element.prototype.matches = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).match(this, selectors);
  };

  Element.prototype.hasAttribute = function (attributeName: string) {
    return nwmatcher({document: xmlDoc}).hasAttribute(this, attributeName);
  };

  Element.prototype.getAttribute = function (attributeName: string) {
    return nwmatcher({document: xmlDoc}).getAttribute(this, attributeName);
  };

  Element.prototype.querySelector = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).first(selectors, this);
  };

  return xmlDoc;
}

export function clearNode(el: Element) {
  Array.from(el.childNodes).forEach(child => child.parentNode!.removeChild(child))
}
