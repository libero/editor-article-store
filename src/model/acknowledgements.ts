import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { makeSchemaFromConfig } from './utils';
import xmldom from 'xmldom';
import {clearNode} from "../xml-exporter/xml-utils";
import {serializeManuscriptSection} from "../xml-exporter/manuscript-serializer";
import {Manuscript} from "../model/manuscript";

export function createAcknowledgementsState(content?: Element | null): EditorState {
  const xmlContentDocument = new xmldom.DOMImplementation().createDocument('', '', null);
  const ack = xmlContentDocument.createElement('ack');
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );

  if (content) {
    ack.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(ack),
    schema
  });
}

export function serializeAcknowledgementState(xmlDoc: Document, manuscript: Manuscript) {
  const acknowledgementsXml = serializeManuscriptSection(manuscript.acknowledgements, xmlDoc);
  let acknowledgementsEl = xmlDoc.querySelector('ack') as Element;

  if (!acknowledgementsEl) {
    acknowledgementsEl = xmlDoc.createElement('ack');
    acknowledgementsEl.setAttribute('id', 'ack');
    const ackTitleEl = xmlDoc.createElement('title');

    ackTitleEl.appendChild(xmlDoc.createTextNode('Acknowledgements'));
    acknowledgementsEl.appendChild(ackTitleEl);
    xmlDoc.querySelector('back')?.appendChild(acknowledgementsEl);
  }

  const ackTitleEl = acknowledgementsEl.firstChild as Element;
  clearNode(acknowledgementsEl);
  acknowledgementsEl.appendChild(ackTitleEl);
  acknowledgementsEl.appendChild(acknowledgementsXml);
}
