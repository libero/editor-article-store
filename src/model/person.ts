import { EditorState } from 'prosemirror-state';
import { get } from 'lodash';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { DOMImplementation } from 'xmldom';

import { BackmatterEntity } from './backmatter-entity';
import * as bioConfig from './config/author-bio.config';
import { JSONObject } from './types';
import { getTextContentFromPath, makeSchemaFromConfig } from './utils';
import { Manuscript } from './manuscript';
import { serializeManuscriptSection } from '../xml-exporter/manuscript-serializer';
import { Affiliation } from './affiliation';
import { clearNode } from '../xml-exporter/xml-utils';

const NO_COI_STATEMENT = 'No competing interests declared';

export class Person extends BackmatterEntity {
    firstName: string | undefined;
    lastName: string | undefined;
    suffix?: string;
    isAuthenticated?: boolean;
    orcid?: string;
    email?: string;
    bio?: EditorState;
    isCorrespondingAuthor?: boolean;
    affiliations?: string[];
    hasCompetingInterest?: boolean;
    competingInterestStatement?: string;

    constructor(data?: JSONObject | Element, notesXml?: Element | null) {
        super();
        this.createEntity(data);
        if (get(data, 'ownerDocument') && get(notesXml, 'ownerDocument')) {
            this.setCompetingInterests(data as Element, notesXml as Element);
        }
    }

    public toXml(authorNotes: Element, affiliations: Affiliation[] = []): Element {
        const xmlDoc = new DOMImplementation().createDocument(null, null);
        const contrib = xmlDoc.createElement('contrib');
        contrib.setAttribute('contrib-type', 'author');
        if (this.isCorrespondingAuthor) {
            contrib.setAttribute('corresp', 'yes');
        }

        const name = xmlDoc.createElement('name');
        contrib.appendChild(name);
        if (this.firstName) {
            const firstName = xmlDoc.createElement('given-names');
            firstName.appendChild(xmlDoc.createTextNode(this.firstName));
            name.appendChild(firstName);
        }

        if (this.lastName) {
            const lastName = xmlDoc.createElement('surname');
            lastName.appendChild(xmlDoc.createTextNode(this.lastName));
            name.appendChild(lastName);
        }

        if (this.suffix) {
            const suffix = xmlDoc.createElement('suffix');
            suffix.appendChild(xmlDoc.createTextNode(this.suffix));
            name.appendChild(suffix);
        }

        if (this.orcid) {
            const orcidEl = xmlDoc.createElement('contrib-id');
            orcidEl.setAttribute('contrib-id-type', 'orcid');
            orcidEl.setAttribute('authenticated', String(this.isAuthenticated));
            orcidEl.appendChild(xmlDoc.createTextNode('https://orcid.org/' + this.orcid));
            contrib.appendChild(orcidEl);
        }

        if (this.email) {
            const emailEl = xmlDoc.createElement('email');
            emailEl.appendChild(xmlDoc.createTextNode(this.email));
            contrib.appendChild(emailEl);
        }

        if (this.bio) {
            const bioXml = xmlDoc.createElement('bio');
            bioXml.appendChild(serializeManuscriptSection(this.bio, xmlDoc));
            contrib.appendChild(bioXml);
        }

        (this.affiliations || []).forEach((affId) => {
            const affEl = xmlDoc.createElement('xref');
            affEl.setAttribute('ref-type', 'aff');
            affEl.setAttribute('rid', affId);
            const affLabel = get(
                affiliations.find((aff) => aff.id === affId),
                'label',
            );
            if (affLabel) {
                affEl.appendChild(xmlDoc.createTextNode(affLabel));
            }
            contrib.appendChild(affEl);
        });
        this.serializeCompetingInterest(contrib, authorNotes);
        return contrib;
    }

    private serializeCompetingInterest(contrib: Element, authorNotes: Element) {
        const xmlDoc = contrib.ownerDocument;
        const coiStatementsList = Array.from(authorNotes.querySelectorAll('fn[fn-type="COI-statement"]'));
        if (this.hasCompetingInterest) {
            let coiStmtEl = coiStatementsList.find((coiEl) => coiEl.textContent === this.competingInterestStatement);

            if (!coiStmtEl) {
                coiStmtEl = xmlDoc.createElement('fn');
                coiStmtEl.setAttribute('fn-type', 'COI-statement');
                coiStmtEl.setAttribute('id', 'con' + (coiStatementsList.length + 1));
                const paragraph = xmlDoc.createElement('p');
                paragraph.appendChild(xmlDoc.createTextNode(this.competingInterestStatement || ''));
                coiStmtEl.appendChild(paragraph);
                authorNotes.appendChild(coiStmtEl);
            }

            const xref = xmlDoc.createElement('xref');
            xref.setAttribute('ref-type', 'author-notes');
            xref.setAttribute('rid', coiStmtEl.getAttribute('id')!);
            contrib.appendChild(xref);
        } else {
            let noCoiStmt = coiStatementsList.find((coiEl) => coiEl.textContent === NO_COI_STATEMENT);
            if (!noCoiStmt) {
                noCoiStmt = xmlDoc.createElement('fn');
                noCoiStmt.setAttribute('fn-type', 'COI-statement');
                noCoiStmt.setAttribute('id', 'con' + (coiStatementsList.length + 1));
                const paragraph = xmlDoc.createElement('p');
                paragraph.appendChild(xmlDoc.createTextNode(NO_COI_STATEMENT));
                noCoiStmt.appendChild(paragraph);
                authorNotes.appendChild(noCoiStmt);
            }

            const xref = xmlDoc.createElement('xref');
            xref.setAttribute('ref-type', 'author-notes');
            xref.setAttribute('rid', noCoiStmt.getAttribute('id')!);
            contrib.appendChild(xref);
        }
    }

    protected fromXML(xml: Element): void {
        const orcidEl = xml.querySelector('contrib-id[contrib-id-type="orcid"]');

        this.firstName = getTextContentFromPath(xml, 'name > given-names');
        this.lastName = getTextContentFromPath(xml, 'name > surname');
        this.suffix = getTextContentFromPath(xml, 'name > suffix');
        this.isAuthenticated = orcidEl ? orcidEl.getAttribute('authenticated') === 'true' : false;
        this.orcid = orcidEl ? this.getOrcid(orcidEl.textContent!) : '';
        this.bio = this.createBioEditorStateFromXml(xml.querySelector('bio')!);
        this.email = getTextContentFromPath(xml, 'email');
        this.isCorrespondingAuthor = xml.getAttribute('corresp') === 'yes';
        this.affiliations = Array.from(xml.querySelectorAll('xref[ref-type="aff"]'))
            .map((xRef) => xRef.getAttribute('rid'))
            .filter(Boolean) as string[];
    }

    protected fromJSON(json: JSONObject): void {
        this._id = (json._id as string) || this.id;
        this.firstName = (json.firstName as string) || '';
        this.lastName = (json.lastName as string) || '';
        this.suffix = (json.suffix as string) || '';
        this.isAuthenticated = !!json.isAuthenticated;
        this.orcid = (json.orcid as string) || '';
        this.bio = json.bio
            ? this.createBioEditorStateFromJSON(json.bio as JSONObject)
            : this.createBioEditorStateFromXml();
        this.email = (json.email as string) || '';
        this.isCorrespondingAuthor = !!json.isCorrespondingAuthor;
        this.hasCompetingInterest = json.hasCompetingInterest as boolean | undefined;
        this.competingInterestStatement = json.competingInterestStatement as string | undefined;
        this.affiliations = Array.isArray(json.affiliations) ? (json.affiliations as string[]) : [];
    }

    protected createBlank(): void {
        this.firstName = '';
        this.lastName = '';
        this.suffix = '';
        this.isAuthenticated = false;
        this.orcid = '';
        this.bio = this.createBioEditorStateFromXml();
        this.email = '';
        this.isCorrespondingAuthor = false;
        this.affiliations = [];
    }

    private setCompetingInterests(dataXml: Element, notesXml: Element): void {
        const competingInterestEl = Array.from(notesXml.querySelectorAll('[fn-type="COI-statement"]')).find(
            (fnEl: Element) => {
                const id = fnEl.getAttribute('id');
                return dataXml.querySelector(`xref[ref-type="author-notes"][rid="${id}"]`);
            },
        );

        this.hasCompetingInterest = competingInterestEl ? competingInterestEl.textContent !== NO_COI_STATEMENT : false;
        this.competingInterestStatement = competingInterestEl ? competingInterestEl.textContent!.trim() : '';
    }

    private createBioEditorStateFromXml(bio?: Element): EditorState {
        const schema = makeSchemaFromConfig(bioConfig.topNode, bioConfig.nodes, bioConfig.marks);
        return EditorState.create({
            doc: bio ? ProseMirrorDOMParser.fromSchema(schema).parse(bio) : undefined,
            schema,
        });
    }

    private createBioEditorStateFromJSON(json: JSONObject): EditorState {
        const blankState = this.createBioEditorStateFromXml();
        return EditorState.fromJSON(
            {
                schema: blankState.schema,
                plugins: blankState.plugins,
            },
            json,
        );
    }

    private getOrcid(orcidUrl: string): string {
        const matches = orcidUrl.match(/(([0-9]{4}-?){4})/g);
        if (matches) {
            return matches[0];
        }
        return '';
    }
}

export function createAuthorsState(authorsXml: Element[], notesXml?: Element | null): Person[] {
    return authorsXml.map((xml) => new Person(xml, notesXml));
}

export function serializeAuthors(xmlDoc: Document, manuscript: Manuscript) {
    let authorsGroup = xmlDoc.querySelector('article-meta > contrib-group');
    if (!authorsGroup) {
        authorsGroup = xmlDoc.createElement('contrib-group');
        xmlDoc.querySelector('article-meta')!.appendChild(authorsGroup);
    } else {
        clearNode(authorsGroup);
    }

    const authorNotes = getAuthorsNotes(xmlDoc);

    manuscript.authors.forEach((author: Person) => {
        authorsGroup!.appendChild(author.toXml(authorNotes, manuscript.affiliations));
    });
}

function getAuthorsNotes(xmlDoc: Document): Element {
    let authorNotes = xmlDoc.documentElement.querySelector('author-notes');
    if (!authorNotes) {
        authorNotes = xmlDoc.createElement('author-notes');
        xmlDoc.querySelector('article-meta')!.appendChild(authorNotes);
    } else {
        authorNotes
            .querySelectorAll('[fn-type="COI-statement"]')
            .forEach((coiStmtEl) => authorNotes!.removeChild(coiStmtEl));
    }

    return authorNotes;
}
