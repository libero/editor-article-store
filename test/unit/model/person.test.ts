import {cloneDeep} from "lodash";
import {EditorState} from "prosemirror-state";

import {createAuthorsState, Person} from "../../../src/model/person";
import {JSONObject} from "../../../src/model/types";
import {clearNode, parseXML} from "../../../src/xml-exporter/xml-utils";

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

const PERSON_JSON_DATA = {
  _id: 'author-3888',
  firstName: 'Siu Sylvia',
  lastName: 'Lee',
  isAuthenticated: true,
  orcid: '0000-0001-5225-4203',
  email: 'sylvia.lee@cornell.edu',
  bio: {
    doc: {
      type: 'doc',
      content: [
        {
          type: 'paragraph',
          content: [
            {type: 'text', marks: [{type: 'bold'}], text: 'Siu Sylvia Lee'},
            {
              type: 'text',
              text:
                ' is in the Department of Molecular Biology and Genetics, Cornell University, Ithaca, United States'
            }
          ]
        }
      ]
    },
    selection: {type: 'text', anchor: 1, head: 1}
  },
  isCorrespondingAuthor: true,
  affiliations: ['aff2']
} as JSONObject;

const PERSON_XML_DATA = `
  <contrib corresp="yes">
    <name><surname>Atherden</surname><given-names>Fred</given-names><suffix>Capt.</suffix></name>
    <contrib-id authenticated="true" contrib-id-type="orcid">https://orcid.org/0000-0002-6048-1470</contrib-id>
    <email>fatherden@elifesciences.org</email>
    <bio><p><bold>Jeanine Smith III</bold> is in the Department, University, City, Country</p></bio>
    <xref ref-type="aff" rid="aff2">2</xref>
    <xref ref-type="aff" rid="aff3">3</xref>
  </contrib>
`;

const AUTHOR_NOTES_XML = `<author-notes>
    <fn fn-type="COI-statement" id="con1">
      <p>Is an employee of eLife. No other competing interests exist</p>
    </fn>
  </author-notes>`;

describe('Person class', () => {

  describe('constructor', function () {
    it('creates a blank Person instance', () => {
      const author = new Person();
      expect(author).toStrictEqual(expect.objectContaining({
        _id: 'unique_id',
        firstName: '',
        lastName: '',
        suffix: '',
        isAuthenticated: false,
        orcid: '',
        email: '',
        isCorrespondingAuthor: false,
        affiliations: [],
      }));

      expect(author.bio).toBeInstanceOf(EditorState);
      expect(author.bio!.doc.textContent).toBe('');
    });
    describe('Person instance from JSON', () => {
      it('creates author from JSON', () => {
        const author = new Person(PERSON_JSON_DATA);
        expect(author).toStrictEqual(expect.objectContaining({
          _id: 'author-3888',
          firstName: 'Siu Sylvia',
          lastName: 'Lee',
          isAuthenticated: true,
          orcid: '0000-0001-5225-4203',
          email: 'sylvia.lee@cornell.edu',
          isCorrespondingAuthor: true,
          affiliations: ['aff2']
        }));

        expect(author.bio!.doc.textContent).toBe('Siu Sylvia Lee is in the Department of Molecular Biology and Genetics, Cornell University, Ithaca, United States')
      });

      it('creates author from JSON and assigns a unique ID if JSON does not have any', () => {
        const jsonData = cloneDeep(PERSON_JSON_DATA);
        jsonData._id = undefined;
        const author = new Person(jsonData);
        expect(author).toStrictEqual(expect.objectContaining({
          _id: 'unique_id',
          firstName: 'Siu Sylvia',
          lastName: 'Lee',
          isAuthenticated: true,
          orcid: '0000-0001-5225-4203',
          email: 'sylvia.lee@cornell.edu',
          isCorrespondingAuthor: true,
          affiliations: ['aff2']
        }));

        expect(author.bio!.doc.textContent).toBe('Siu Sylvia Lee is in the Department of Molecular Biology and Genetics, Cornell University, Ithaca, United States')
      });

      it('ignores non-array values of affiliations field ', () => {
        const author = new Person({affiliations: {testKey: 'testValue'}});
        expect(author).toStrictEqual(expect.objectContaining({
          _id: 'unique_id',
          firstName: '',
          lastName: '',
          isAuthenticated: false,
          orcid: '',
          email: '',
          isCorrespondingAuthor: false,
          affiliations: []
        }));

        expect(author.bio!.doc.textContent).toBe('')
      });
    });

    describe('Person instance from XML', () => {
      it('creates author from XML', () => {

        const authorXml = parseXML(PERSON_XML_DATA);

        const author = new Person(authorXml.querySelector('contrib')!);
        expect(author).toStrictEqual(expect.objectContaining({
          _id: 'unique_id',
          firstName: 'Fred',
          lastName: 'Atherden',
          isAuthenticated: true,
          orcid: '0000-0002-6048-1470',
          email: 'fatherden@elifesciences.org',
          isCorrespondingAuthor: true,
          affiliations: ['aff2', 'aff3']
        }));

        expect(author.bio!.doc.textContent).toBe('Jeanine Smith III is in the Department, University, City, Country')
      });

      it('correctly assigns the authenticated flag attribute from XML', () => {
        const authorXml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;

        expect(new Person(authorXml).isAuthenticated).toBe(true);

        authorXml.querySelector('contrib-id')!.setAttribute('authenticated', 'gibberish');
        expect(new Person(authorXml).isAuthenticated).toBe(false);

        authorXml.querySelector('contrib-id')!.removeAttribute('authenticated');
        expect(new Person(authorXml).isAuthenticated).toBe(false);

        const orcidEl = authorXml.querySelector('contrib-id');
        orcidEl!.parentNode!.removeChild(orcidEl!);
        expect(new Person(authorXml).isAuthenticated).toBe(false);
      });

      it('correctly assigns orcid from XML', () => {
        const authorXml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;

        expect(new Person(authorXml).orcid).toBe('0000-0002-6048-1470');

        setNodeTextContent(
          authorXml.querySelector('contrib-id')!,
          'https://www.google.com/search?q=codswallop'
        );
        expect(new Person(authorXml).orcid).toBe('');

        setNodeTextContent(authorXml.querySelector('contrib-id')!, '');
        expect(new Person(authorXml).orcid).toBe('');

        const orcidEl = authorXml.querySelector('contrib-id');
        orcidEl!.parentNode!.removeChild(orcidEl!);
        expect(new Person(authorXml).orcid).toBe('');
      });

      it('correctly assigns corresponding author flag from XML', () => {
        const authorXml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;

        expect(new Person(authorXml).isCorrespondingAuthor).toBe(true);

        authorXml.setAttribute('corresp', 'yas');
        expect(new Person(authorXml).isCorrespondingAuthor).toBe(false);

        authorXml.removeAttribute('corresp');
        expect(new Person(authorXml).isCorrespondingAuthor).toBe(false);
      });

      it('correctly parses bio from XML', () => {
        const authorXml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;

        expect(new Person(authorXml).bio!.doc.textContent)
          .toBe('Jeanine Smith III is in the Department, University, City, Country');

        const bioEl = authorXml.querySelector('bio');
        bioEl!.parentNode!.removeChild(bioEl!);
        expect(new Person(authorXml).bio!.doc.textContent).toBe('');
      });

      it('correctly parses linked affiliations from XML', () => {
        const authorXml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;

        const linkedAffiliations = authorXml.querySelectorAll('xref[ref-type="aff"]');
        expect(new Person(authorXml).affiliations).toEqual(['aff2', 'aff3']);

        authorXml.removeChild(linkedAffiliations[0]);
        expect(new Person(authorXml).affiliations).toEqual(['aff3']);

        authorXml.removeChild(linkedAffiliations[1]);
        expect(new Person(authorXml).affiliations).toEqual([]);
      });
    });

    describe('creates author with competing interests', () => {
      it('creates an author with competing interests', () => {
        const notesXml = parseXML(AUTHOR_NOTES_XML).querySelector('author-notes')!;
        const authorXml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;

        const author1 = new Person(authorXml, notesXml);
        expect(author1.hasCompetingInterest).toBe(false);
        expect(author1.competingInterestStatement).toBe('');

        const competingInterestRef = authorXml.ownerDocument.createElement('xref');
        competingInterestRef.setAttribute('ref-type', 'fn');
        competingInterestRef.setAttribute('rid', 'con1');
        authorXml.appendChild(competingInterestRef);

        const author2 = new Person(authorXml, notesXml);
        expect(author2.hasCompetingInterest).toBe(true);
        expect(author2.competingInterestStatement).toBe('Is an employee of eLife. No other competing interests exist');
      });
    });
  });

  describe('creates authors state', () => {
    it('creates an empty authors state', () => {
      expect(createAuthorsState([])).toEqual([]);
    });

    it('creates an empty authors state without notes', () => {
      const author1Xml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;
      author1Xml.setAttribute('id', 'author-3888');

      const author2Xml = parseXML(`<contrib corresp="yes">
        <name><surname>Lee</surname><given-names>Siu Sylvia</given-names></name>
        <contrib-id authenticated="true" contrib-id-type="orcid">https://orcid.org0000-0001-5225-4203</contrib-id>
        <email>sylvia.lee@cornell.edu</email>
        <xref ref-type="aff" rid="aff2">2</xref>
      </contrib>`).querySelector('contrib')!;
      author2Xml.setAttribute('id', 'author-3889');

      const state = createAuthorsState([author1Xml, author2Xml]);
      expect(state.length).toBe(2);
      expect(state[0]).toStrictEqual(expect.objectContaining({
        _id: 'author-3888',
        firstName: 'Fred',
        lastName: 'Atherden',
        isAuthenticated: true,
        orcid: '0000-0002-6048-1470',
        email: 'fatherden@elifesciences.org',
        isCorrespondingAuthor: true,
        affiliations: ['aff2', 'aff3']
      }));

      expect(state[1]).toStrictEqual(expect.objectContaining({
        _id: 'author-3889',
        firstName: 'Siu Sylvia',
        lastName: 'Lee',
        isAuthenticated: true,
        orcid: '0000-0001-5225-4203',
        email: 'sylvia.lee@cornell.edu',
        isCorrespondingAuthor: true,
        affiliations: ['aff2']
      }));
    });

    it('creates an empty authors state with notes', () => {
      const author1Xml = parseXML(PERSON_XML_DATA).querySelector('contrib')!;
      author1Xml.setAttribute('id', 'author-3888');

      const competingInterestRef = author1Xml.ownerDocument.createElement('xref');
      competingInterestRef.setAttribute('ref-type', 'fn');
      competingInterestRef.setAttribute('rid', 'con1');
      author1Xml.appendChild(competingInterestRef);

      const author2Xml = parseXML(`<contrib corresp="yes">
        <name><surname>Lee</surname><given-names>Siu Sylvia</given-names></name>
        <contrib-id authenticated="true" contrib-id-type="orcid">https://orcid.org0000-0001-5225-4203</contrib-id>
        <email>sylvia.lee@cornell.edu</email>
        <xref ref-type="aff" rid="aff2">2</xref>
      </contrib>`).querySelector('contrib')!;
      author2Xml.setAttribute('id', 'author-3889');

      const notesXml = parseXML(AUTHOR_NOTES_XML).querySelector('author-notes')!;

      const state = createAuthorsState([author1Xml, author2Xml], notesXml);
      expect(state.length).toBe(2);
      expect(state[0]).toStrictEqual(expect.objectContaining({
        _id: 'author-3888',
        firstName: 'Fred',
        lastName: 'Atherden',
        suffix: "Capt.",
        isAuthenticated: true,
        hasCompetingInterest: true,
        competingInterestStatement: 'Is an employee of eLife. No other competing interests exist',
        orcid: '0000-0002-6048-1470',
        email: 'fatherden@elifesciences.org',
        isCorrespondingAuthor: true,
        affiliations: ['aff2', 'aff3']
      }));

      expect(state[1]).toStrictEqual(expect.objectContaining({
        _id: 'author-3889',
        firstName: 'Siu Sylvia',
        lastName: 'Lee',
        isAuthenticated: true,
        hasCompetingInterest: false,
        competingInterestStatement: '',
        orcid: '0000-0001-5225-4203',
        email: 'sylvia.lee@cornell.edu',
        isCorrespondingAuthor: true,
        affiliations: ['aff2']
      }));
    });
  });
});

function setNodeTextContent(xmlNode: Element, text: string) {
  clearNode(xmlNode);
  xmlNode.appendChild(xmlNode.ownerDocument.createTextNode(text));
}
