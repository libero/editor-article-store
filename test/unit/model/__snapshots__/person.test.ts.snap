// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person class Serialization to XML serializes a Person object to XML 1`] = `"<article><article-meta><contrib-group><contrib contrib-type=\\"author\\" id=\\"author-3888\\" corresp=\\"yes\\"><name><given-names>Siu Sylvia</given-names><surname>Lee</surname></name><contrib-id contrib-id-type=\\"orcid\\" authenticated=\\"true\\">https://orcid.org/0000-0001-5225-4203</contrib-id><email>sylvia.lee@cornell.edu</email><bio><p><bold>Siu Sylvia Lee</bold> is in the Department of Molecular Biology and Genetics, Cornell University, Ithaca, United States</p></bio></contrib></contrib-group></article-meta></article>"`;
